sim/midas/src/main/scala/midas/passes/fame/FindDefaultClocks.scala

```diff
@@ -6,46 +6,96 @@ import firrtl._
 
-  def recordDefaultClocks(topTarget: ModuleTarget, defaultClocks: ClockConnMap)(stmt: Statement): Unit = stmt match {
+  /**
+    * By looking at the top-level connects on ClockTypes, build a graph of
+    * connectivity that will allow us to tie cliques of models back to a source
+    * port on the hub.
+    */
+  def recordClockConnection(
+      topTarget: ModuleTarget,
+      defaultClocks: MutableDiGraph[ClockConnNode],
+      hubModel: String,
+      )(stmt: Statement): Unit = stmt match {
     case Connect(_, WSubField(WRef(lInst, _, InstanceKind, _), lPort, ClockType, _), WSubField(WRef(rInst, _, InstanceKind, _), rPort, ClockType, _)) =>
-      // LHS must be "satellite" model receiving clock from "hub" model on RHS
-      defaultClocks(ModelInstance(lInst)) = ClockConnection(topTarget.ref(rInst).field(rPort), topTarget.ref(lInst).field(lPort))
-    case s => s.foreach(recordDefaultClocks(topTarget, defaultClocks))
+        defaultClocks.addPairWithEdge(
+          ClockConnNode(lInst, if (lInst == hubModel) Some(topTarget.ref(lInst).field(lPort)) else None),
+          ClockConnNode(rInst, if (rInst == hubModel) Some(topTarget.ref(rInst).field(rPort)) else None))
+    case s => s.foreach(recordClockConnection(topTarget, defaultClocks, hubModel))
+  }
+
+  // Finds a driver (a source port on the hub model) for a given model instance
+  def getDefaultClock(clockGraph: ClockConnMap, satModelInstName: String): ReferenceTarget = {
+    clockGraph.reachableFrom(ClockConnNode(satModelInstName)).collectFirst {
+      case ClockConnNode(_, Some(port)) => port
+    }.get
   }
 
-  def addDefaultClocks(hubModel: ModelInstance, defaultClocks: ClockConnMap)(anno: Annotation): Annotation = anno match {
+
+  def addDefaultClocks(hubName: String, clockGraph: ClockConnMap)(anno: Annotation): Annotation = anno match {
     case fcca @ FAMEChannelConnectionAnnotation(name, info, None, Some(sources), Some(sinks)) =>
       // unclocked loopback channel
-      val sourceModelInst = ModelInstance(sources.head.ref)
-      val sinkModelInst = ModelInstance(sinks.head.ref)
-      if (sourceModelInst == hubModel) {
-        fcca.copy(clock = Some(defaultClocks(sinkModelInst).source))
-      } else if (sinkModelInst == hubModel) {
-        fcca.copy(clock = Some(defaultClocks(sourceModelInst).source))
+      val sourceModelInst = sources.head.ref
+      val sinkModelInst = sinks.head.ref
+      if (sinkModelInst != hubName) {
+        fcca.copy(clock = Some(getDefaultClock(clockGraph, sinkModelInst)))
+      } else if (sourceModelInst != hubName) {
+        fcca.copy(clock = Some(getDefaultClock(clockGraph, sourceModelInst)))
       } else {
         fcca
       }
     case a => a
   }
 
+  // Hub self-loops can be introduced as a result of extracting passthroughs in
+  // satellite models.  For these channels, a default clock cannot be inferred by
+  // looking at Connect statements.  Instead find a second channel that shares the
```

Comment:
Is this always guaranteed to exist?

Reply:
The way i've currently written this, the model from which the passthrough was extracted should remain a sink in a fanout inclusive of this self-loop, even if it's unused the satellite model.